### ISSUE TITLE
[Research] Add Cross-Modal EEG+Voice Temporal Alignment with Optimal Transport Fusion

### ISSUE THREAD

**Issue by LakshmiSravyaVedantham** [LakshmiSravyaVedantham/DreamAnalysis#55]
## Papers

**Primary:** Multimodal Emotion Recognition via the Fusion of Mamba and Liquid Neural Networks with Cross-Modal Alignment
**Authors:** Various
**Year:** 2024
**Venue:** Electronics (MDPI)
**URL:** https://www.mdpi.com/2079-9292/14/18/3638

**Supporting:** Cross-modal alignment and fusion of EEG-visual based on mixed attention mechanism for emotion recognition
**Year:** 2025
**Venue:** Cognitive Neurodynamics (Springer)
**URL:** https://link.springer.com/article/10.1007/s11571-025-10372-5

**Supporting:** A Comprehensive Review of Multimodal Emotion Recognition: Techniques, Challenges, and Future Directions (PMC, 2025)

## Key Contribution

Introduces optimal transport (OT) algorithms for aligning features across modalities (EEG, speech, video, text) that operate at different temporal resolutions. The Mamba+Liquid Neural Network architecture captures neural dynamics through high-resolution EEG spectrograms (STFT) while aligning with other modality features. Over 40% of 2022+ studies adopt transformer-style cross-modal fusion, consistently outperforming unimodal approaches by 5-15%.

The cross-modal attention paper (2025) specifically addresses the challenge that EEG features are extracted at ~256Hz while speech features come at ~16kHz (raw) or ~100Hz (frame-level), requiring explicit temporal alignment before fusion.

## How It Applies to NeuralDreamWorkshop

NeuralDreamWorkshop already has both EEG (Muse 2) and voice emotion (emotion2vec/SenseVoice) modalities, but they operate independently:
- EEG: 256 Hz, 4-second epochs, emotion every 2 seconds
- Voice: processed per utterance (~1-10 seconds), emotion per segment

Issue #21 proposed TMNet for EEG+Voice fusion but didn't address the temporal alignment problem. This research provides the missing synchronization layer.

Key challenge: when a user speaks while wearing the Muse 2, the voice emotion and EEG emotion predictions must be temporally aligned to the same emotional moment before fusion can work.

### Files to Modify/Create
- `ml/processing/temporal_aligner.py` — new module for cross-modal alignment
- `ml/models/multimodal_fusion.py` — lightweight fusion model (extend TMNet from #21)
- `ml/api/routes.py` — add `/analyze-multimodal` endpoint that accepts both EEG + audio

### Core Algorithm
```python
import numpy as np
from scipy.interpolate import interp1d

class TemporalAligner:
    """
    Aligns EEG and voice emotion features to a common timebase.
    
    EEG: features every 2 seconds (4-sec epoch, 50% overlap)
    Voice: features per utterance (variable length, 1-10 sec)
    
    Uses linear interpolation + attention-weighted fusion.
    """
    
    COMMON_RATE_HZ = 0.5  # 1 prediction every 2 seconds
    
    def align_eeg_voice(
        self,
        eeg_features: list[dict],   # [{timestamp, valence, arousal, ...}, ...]
        voice_features: list[dict],  # [{start_time, end_time, emotion, probs, ...}, ...]
        session_start: float,
    ) -> list[dict]:
        """
        Align EEG and voice features to common 2-second grid.
        Returns fused predictions at each timepoint.
        """
        # 1. Build common time grid
        duration = max(
            max(e['timestamp'] for e in eeg_features),
            max(v['end_time'] for v in voice_features)
        ) - session_start
        time_grid = np.arange(0, duration, 1.0 / self.COMMON_RATE_HZ)
        
        # 2. Interpolate EEG features to grid
        eeg_on_grid = self._interpolate_to_grid(eeg_features, time_grid, 'timestamp')
        
        # 3. Spread voice features to grid (utterance covers a time span)
        voice_on_grid = self._spread_utterances_to_grid(voice_features, time_grid)
        
        # 4. Confidence-weighted fusion at each timepoint
        fused = []
        for t, eeg_f, voice_f in zip(time_grid, eeg_on_grid, voice_on_grid):
            fused.append(self._fuse_timepoint(t, eeg_f, voice_f))
        
        return fused
    
    def _fuse_timepoint(self, t, eeg_f, voice_f):
        """
        Fuse EEG and voice predictions at a single timepoint.
        
        Strategy: dynamic weighting based on signal quality.
        - If voice is present and confident → weight voice higher (voice is better for discrete emotions)
        - If only EEG → use EEG alone
        - Valence: 60% EEG (FAA) + 40% voice (if available)
        - Arousal: 50/50 blend (EEG beta/alpha + voice energy/pitch)
        - Discrete emotion: 40% EEG + 60% voice (voice is more discriminative)
        """
        if voice_f is None or voice_f.get('confidence', 0) < 0.3:
            # No voice or low confidence — EEG only
            return {**eeg_f, 'modality': 'eeg_only', 'timestamp': t}
        
        if eeg_f is None or eeg_f.get('epoch_ready', True) is False:
            # No EEG epoch ready — voice only
            return {**voice_f, 'modality': 'voice_only', 'timestamp': t}
        
        # Fused prediction
        eeg_w_valence, voice_w_valence = 0.60, 0.40
        eeg_w_arousal, voice_w_arousal = 0.50, 0.50
        eeg_w_emotion, voice_w_emotion = 0.40, 0.60
        
        fused_valence = (eeg_w_valence * eeg_f['valence'] + 
                         voice_w_valence * voice_f.get('valence', 0))
        fused_arousal = (eeg_w_arousal * eeg_f['arousal'] + 
                         voice_w_arousal * voice_f.get('arousal', 0.5))
        
        # Fuse discrete emotion probabilities
        fused_probs = {}
        for emotion in ['happy', 'sad', 'angry', 'fear', 'surprise', 'neutral']:
            eeg_p = eeg_f.get('probabilities', {}).get(emotion, 1/6)
            voice_p = voice_f.get('probabilities', {}).get(emotion, 1/6)
            fused_probs[emotion] = eeg_w_emotion * eeg_p + voice_w_emotion * voice_p
        
        # Normalize
        total = sum(fused_probs.values())
        fused_probs = {k: v/total for k, v in fused_probs.items()}
        
        return {
            'timestamp': t,
            'emotion': max(fused_probs, key=fused_probs.get),
            'probabilities': fused_probs,
            'valence': float(fused_valence),
            'arousal': float(fused_arousal),
            'modality': 'eeg+voice',
            'eeg_weight': eeg_w_emotion,
            'voice_weight': voice_w_emotion,
        }
```

### Synchronization Protocol
```
Browser captures:
  1. EEG stream via Web Bluetooth → /analyze-eeg (every 2 sec)
  2. Audio stream via MediaRecorder → /analyze-voice (per utterance)
  3. Both tagged with performance.now() timestamp
  
Server aligns:
  - Common session clock established at session start
  - TemporalAligner.align_eeg_voice() called periodically or at session end
  - Real-time fusion: latest EEG + latest voice within 2-sec window
```

## Expected Performance Impact

- Multimodal fusion consistently adds +5-15% over best unimodal (from survey of 40%+ of studies)
- Voice is significantly better than EEG for discrete emotions (happy vs angry vs sad): ~75-85% vs 55-65%
- EEG is better for continuous dimensions (valence trend, arousal level)
- Expected fused accuracy: 70-80% discrete emotion (vs 55-65% EEG-only)
- Main value: voice fills in the gaps when EEG is noisy/artifact-heavy, and EEG provides continuous monitoring between speech segments

## Complexity: MEDIUM

- Temporal alignment is straightforward (interpolation + grid mapping)
- The main engineering challenge is synchronizing browser timestamps between Web Bluetooth (EEG) and MediaRecorder (audio) — both use different clocks
- No new model training for v1 (weighted average fusion)
- v2 could add learned attention weights (requires labeled multimodal data)
- Builds on existing infrastructure: EEG pipeline + voice-watch already operational

## Dependencies
- Voice emotion pipeline (emotion2vec or SenseVoice — issues #23, #36)
- EEG pipeline (already operational)
- Browser-side timestamp synchronization (client-side change)
- Extends issue #21 (TMNet EEG+Voice Fusion) with the temporal alignment solution
